File: src/main/java/pocxor/XorPlotter.java
package pocxor;

import brs.util.Convert;
import brs.util.MiningPlot;

import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class XorPlotter {

    private final long id;

    public XorPlotter(long id) {
        this.id = id;
    }

    public void plot() {
        try (FileOutputStream file = new FileOutputStream(Convert.toUnsignedLong(id), false)) {
            byte[] data = calculatePlotData();

            file.write(data);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private byte[] calculatePlotData() {
        byte data[] = new byte[MiningPlot.SCOOPS_PER_PLOT * MiningPlot.PLOT_SIZE];

        for (long nonce = 0; nonce < MiningPlot.SCOOPS_PER_PLOT; nonce++) { // first set is straight copied
            MiningPlot plot = new MiningPlot(id, nonce);
            for (int scoop = 0; scoop < MiningPlot.SCOOPS_PER_PLOT; scoop++) {
                System.arraycopy(plot.getScoop(scoop), 0, data, (int) ((nonce * MiningPlot.PLOT_SIZE) + (scoop * MiningPlot.SCOOP_SIZE)), MiningPlot.SCOOP_SIZE);
            }
        }

        System.out.println("starting 2nd");

        for (long nonce = 0; nonce < MiningPlot.SCOOPS_PER_PLOT; nonce++) { // second set is nonce + SCOOPS_PER_PLOT and is xored
            MiningPlot plot = new MiningPlot(id, nonce + MiningPlot.SCOOPS_PER_PLOT);
            for (int scoop = 0; scoop < MiningPlot.SCOOPS_PER_PLOT; scoop++) {
                XorUtil.xorArray(data, (int) ((nonce * MiningPlot.SCOOP_SIZE) + (scoop * MiningPlot.PLOT_SIZE)), plot.getScoop(scoop));
            }
        }

        return data;
    }
}

File: src/main/kotlin/burst/miner/pocxor/XorPlotter.kt
package burst.miner.pocxor

import burst.kit.crypto.BurstCrypto

import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.util.function.Supplier

class XorPlotter(private val id: Long, private val startNonce: Long) {
    private var burstCrypto = BurstCrypto.getInstance()

    fun plot() {
        try {
            FileOutputStream(java.lang.Long.toUnsignedString(id) + "_" + startNonce.toString(), false).use { file ->
                val data = calculatePlotData()
                file.write(data)
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun calculatePlotData(): ByteArray {
        val data = ByteArray(MiningPlot.SCOOPS_PER_PLOT * MiningPlot.PLOT_SIZE)

        for (nonce in 0 until MiningPlot.SCOOPS_PER_PLOT) { // first set is straight copied
            println(nonce + startNonce)
            val plot = MiningPlot(Supplier { burstCrypto.shabal256 }, id, nonce + startNonce, 2)
            for (scoop in 0 until MiningPlot.SCOOPS_PER_PLOT) {
                System.arraycopy(plot.getScoop(scoop), 0, data, nonce * MiningPlot.PLOT_SIZE + scoop * MiningPlot.SCOOP_SIZE, MiningPlot.SCOOP_SIZE)
            }
        }

        println("starting 2nd")

        for (nonce in 0 until MiningPlot.SCOOPS_PER_PLOT) { // second set is nonce + SCOOPS_PER_PLOT and is xored
            println(nonce + MiningPlot.SCOOPS_PER_PLOT + startNonce)
            val plot = MiningPlot(Supplier { burstCrypto.shabal256 }, id, nonce + MiningPlot.SCOOPS_PER_PLOT + startNonce, 2)
            for (scoop in 0 until MiningPlot.SCOOPS_PER_PLOT) {
                XorUtil.xorArray(data, (nonce * MiningPlot.SCOOP_SIZE + scoop * MiningPlot.PLOT_SIZE), plot.getScoop(scoop))
            }
        }

        return data
    }
}
